R/thumbnail.R

```diff
@@ -0,0 +1,327 @@
+#' no path is provided, a temporary file with the correct file extension is
+#' created. If the content item does not have a thumbnail, returns `NA`.
+#' 
+#' @returns The path to the image file.
+#' 
+#' @examples
+#' \dontrun{
+#' client <- connect()
+#' item <- content_item(client, "8f37d6e0-3395-4a2c-aa6a-d7f2fe1babd0")
+#' thumbnail <- get_thumbnail(item)
+#' }
+#'
+#' @family thumbnail functions
+#' @family content functions
+#' @export
+get_thumbnail <- function(content, path = NULL) {
+  validate_R6_class(content, "Content")
+  guid <- content$content$guid
+  con <- content$connect
+
+  # Connect 2024.09.0 introduced public endpoints for content thumbnails. We
+  # prefer those, falling back to the unversioned endpoints if unavailable.
+  res <- con$GET(
+    url = con$server_url("content", guid, "__thumbnail__"),
+    parser = NULL
+  )
+  if (httr::status_code(res) == 404) {
+    res <- con$GET(
+      unversioned_url("applications", guid, "image"),
+      parser = NULL
+    )
+  }
+  con$raise_error(res)
+
+  if (httr::status_code(res) == 204) {
+    return(NA_character_)
+  }
+
+  # Guess file extension
+  if (is.null(path)) {
+    ct <- httr::headers(res)$`content-type`
+    if (grepl("^image/", ct)) {
+      # Just strip off 'image/'
+      ext <- names(mime::mimemap[mime::mimemap == ct])[1]
+      path <- tempfile(pattern = "content_image_", fileext = paste0(".", ext))
+    } else {
+      # Try png
+      warning(glue::glue("Could not infer file extension from content type: {ct}. Using '.png'"))
+      path <- tempfile(pattern = "content_image_", fileext = ".png")
+    }
+  }
+
+  writeBin(httr::content(res, as = "raw"), path)
+
+  return(path)
+}
+
+
+#' Delete content item thumbnail
+#'
+#' Delete the thumbnail from a content item on Connect.
+#'
+#' @param content A content item.
+#' 
+#' @returns The content item (invisibly).
+#' 
+#' @examples
+#' \dontrun{
+#' client <- connect()
+#' item <- content_item(client, "8f37d6e0-3395-4a2c-aa6a-d7f2fe1babd0")
+#' thumbnail <- get_thumbnail(item)
+#' }
+#'
+#' @family thumbnail functions
+#' @family content functions
+#' @export
+delete_thumbnail <- function(content) {
+  validate_R6_class(content, "Content")
+  guid <- content$content$guid
+  con <- content$connect
+
+  # Connect 2024.09.0 introduced public endpoints for content thumbnails. We
+  # prefer those, falling back to the unversioned endpoints if unavailable.
+  res <- con$DELETE(
+    v1_url("content", guid, "thumbnail"),
+    parser = NULL
+  )
+  if (httr::status_code(res) == 404) {
+    res <- con$DELETE(
+      unversioned_url("applications", guid, "image"),
+      parser = NULL
+    )
+  }
+  con$raise_error(res)
+
+  invisible(content)
+}
+
+#' Check content item thumbnail
+#'
+#' Check whether a content item has a thumbnail.
+#'
+#' @param content A content item.
+#' 
+#' @returns `TRUE` if the content item has a thumbnail, otherwise `FALSE`.
```

Comment:
note: the true/false are accurate only when the client has permissions to view the thumbnail.